### PR TITLE
Fan no off cycle action

### PR DESCRIPTION
**Pull request by skykingjwc**
# What does this implement/fix?
Adds an option to the fan cycle_speed action so that the fan will not turn off when calling the cycle_speed action if so desired by the user

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3283

## Test Environment

- [X] ESP32
- [X] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
esphome:
    name: fantest

esp32:
  board: esp32dev
  framework:
    type: esp-idf

output:
  - platform: ledc
    id: my_output_id
    inverted: false
    pin: GPIO19

logger:

fan:
    - platform: speed
      id: test_speed_fan
      speed_count: 3
     
      output: my_output_id
      on_speed_set:
        then:
          - logger.log: Fan speed was changed!

binary_sensor:
  - platform: gpio
    name: Cycle Fan Speed
    id: cyclespeed
    pin:
      number: 18
      mode:
        input: true
        pulldown: true

    filters:
      - delayed_on: 10ms

    # Optional variables:
    on_press:
      - fan.cycle_speed: 
          id: test_speed_fan
          #off_speed_cycle: False
      - logger.log: "Cycle speed clicked"
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by skykingjwc** [comment]
This is associated with PR https://github.com/esphome/esphome/pull/5550   
If this PR is accepted, should probably delete PR 5550.

File: esphome/components/fan/__init__.py [hunk 1]
```diff
@@ -172,6 +173,7 @@ async def create_fan_state(config):
 FAN_ACTION_SCHEMA = maybe_simple_id(
     {
         cv.Required(CONF_ID): cv.use_id(Fan),
+        cv.Optional(CONF_OFF_SPEED_CYCLE, default=True): cv.boolean,
```
Comment (jesserockz) [inline esphome/components/fan/__init__.py]:
```suggestion
```

File: esphome/components/fan/__init__.py [hunk 2]
```diff
@@ -220,7 +222,10 @@ async def fan_turn_on_to_code(config, action_id, template_arg, args):
 @automation.register_action("fan.cycle_speed", CycleSpeedAction, FAN_ACTION_SCHEMA)
```
Comment (jesserockz) [inline esphome/components/fan/__init__.py]: Thois should just extend the action schema as its only used in this action
```suggestion
@automation.register_action(
    "fan.cycle_speed", 
    CycleSpeedAction, 
    FAN_ACTION_SCHEMA.extend({
        cv.Optional(CONF_OFF_SPEED_CYCLE, default=True): cv.boolean,
    }),
)
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Much cleaner, Thanks for adjusting.
Just need to clean up the schema as per below.

**Review by jesserockz — DISMISSED** [review]

**Comment by skykingjwc** [comment]
FYI after the last commit, now getting an error during compile.
```
esphome\esphome\components\fan\__init__.py", line 224, in <module>
    FAN_ACTION_SCHEMA.extend(
AttributeError: 'function' object has no attribute 'extend
```

File: esphome/components/fan/__init__.py [hunk 3]
```diff
@@ -217,10 +218,21 @@ async def fan_turn_on_to_code(config, action_id, template_arg, args):
...
 
-@automation.register_action("fan.cycle_speed", CycleSpeedAction, FAN_ACTION_SCHEMA)
+@automation.register_action(
+    "fan.cycle_speed",
+    CycleSpeedAction,
+    FAN_ACTION_SCHEMA.extend(
+        {
+            cv.Optional(CONF_OFF_SPEED_CYCLE, default=True): cv.boolean,
+        }
+    ),
```
Comment (jesserockz) [inline esphome/components/fan/__init__.py]: In that case, we will just reimplement, not extend
```suggestion
    maybe_simple_id(
        {
            cv.Required(CONF_ID): cv.use_id(Fan),
            cv.Optional(CONF_OFF_SPEED_CYCLE, default=True): cv.boolean,
        }
    ),
```

Comment (skykingjwc) [reply esphome/components/fan/__init__.py]: Thanks.  I'm primarily a web developer, and don't know all the features of C++.  Sometimes I don't see the obvious solution, and this project is one of the most complex things I've ever seen.  I appreciate your consideration.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**